### ISSUE TITLE
Missing documentation for using the Dataset API in combination with image summaries

### ISSUE THREAD

**Issue by stecklin** [tensorflow/tensorflow#15198]
The `Dataset` API is now the recommended input pipeline, however I am missing some guidance on how to include summaries of my images.

```python
def get_data():
  dataset = FixedLengthRecordDataset(...)
  dataset = dataset.map(parse_dataset, ...)
  if is_training:
    dataset = dataset.map(preprocess_for_train, ...)
  # Do shuffling, batching...
  return dataset

def preprocess_for_train(image, label):
  # Do preprocessing...
  image = tf.image.random_flip_left_right(image)
  # Add summary
  tf.summary.image('preprocessed_image', tf.expand_dims(image, 0))
  return image, label
```

This is what I would do intuitively, but since `map()` uses a different thread and therefore a different `tf.Graph` instance (?), the summaries are lost.

What is the recommended way of adding image summaries when using the `Dataset` API? I would like to request a comment / example on that in the official docs.

**Comment by marcoadurno** [comment]
Hi,

afaik you can run the summary, but only after you created the iterator and fetched the next element:
```python
dataset = get_data(is_training=True)
t_example = dataset.make_one_shot_iterator().get_next()
tf.summary.image('preprocessed_image', t_example['image'])
```

**Comment by stecklin** [comment]
@marcoadurno That's true, but it comes with a drawback. Assume the training loop looks like this:

```python
image_batch, label_batch = sess.run(t_example)
_, summary = sess.run([train_op, summary_op], 
                      feed_dict={image: image_batch, label: label_batch})
```

This will invoke the iterator's `get_next` twice, once you run `t_example` and once you request the summary since you introduced a dependency.

**Comment by marcoadurno** [comment]
@stecklin I'm not sure I fully understand the reason why image_batch and label_batch have be be fetched using sess.run to then fed back in using placeholders. If doing so, I think the `summary_op` should use the `image` placeholder, same as `train_op` does but I struggle to see why mixing the Data api with placeholders.

If things can be shuffled around:
```python
dataset = get_data(is_training=True)
t_example = dataset.make_one_shot_iterator().get_next()
t_logits = infer_fn(t_example['image'])
loss_op = loss_fn(t_example['labels'], t_logits)
train_op = train_fn(loss_op)
summary_op = tf.summary.image('preprocessed_image', t_example['image'])
for i in xrange(n_steps):
  fetches = [loss_op, train_op]
  if i%100:
    fetches += [summary_op]
  sess.run(fetches)
```
Or alternatively you can use a monitored session and specify summary hooks. But the basic idea is that if you have a canonical training loop there is no reason why summaries can't be written using the Data api.

**Comment by drpngx** [comment]
@MarkDaoust maybe we need more docs.

/CC @mrry

**Comment by stecklin** [comment]
@marcoadurno Thank you so much for pointing that out. You made me realize that I am not using the `Dataset API` as it is meant to be used, that there is no need for placeholders. 
Summaries can be added as you suggested, works like a charm.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by MarkDaoust** [comment]
Thanks for sorting that out @marcoadurno. This looks like the right solution to me.

Yes, AFAIK, you can't recover summaries from inside a `map` (hopefully @mrry will correct me if I'm wrong). I'll add a note to the `Dataset` doc explaining this.

Note, in the future, stackoverflow might be a better place for this sort of question.
